# Tuning an XGBoost Model

The goal of this notebook is to train and evaluate an XGBoost model, comparing it's performance on a holdout set against other types of models (LR, SVC, LDA). 

To ensure reproducibility and consistent evaluation across models, all datasets were **pre-split into cross-val data and holdout data** as below:

| Split type           | CV training file     | Holdout file              | Description                              |
| -------------------- | -------------------- | ------------------------- | ---------------------------------------- |
| **Random**           | `apps_cv_random.csv` | `apps_holdout_random.csv` | Simple random sampling                   |
| **Stratified**       | `apps_cv_strat.csv`  | `apps_holdout_strat.csv`  | Stratified by `TARGET`                   |
| **Multi-Stratified** | `apps_cv_multi.csv`  | `apps_holdout_multi.csv`  | Stratified by `TARGET` + `CODE_GENDER_M` |

Each dataset for cross-validation (`apps_cv_*.csv`) also contains a column, `fold`, with pre-assigned folds from 1-5 using the corresponding splitting method to ensure consistent evaluation. Therefore, no additional splitting is needed inside this notebook -- can simply loop through assigned folds for cross-validation.


In [144]:
import pandas as pd 
import numpy as np 
from xgboost import XGBClassifier
from itertools import product
import time

## Evaluation Functions

#### Metric calculators:

Copied from `cross_val.ipynb`

In [145]:
# METRICS 

def classification_metrics(y_true, y_pred):
    """
    Computes confusion matrix + accuracy, precision, recall, F1, and balanced accuracy.
    """
    y_true = np.asarray(y_true)
    y_pred = np.asarray(y_pred)

    # Confusion matrix components
    tp = np.sum((y_true == 1) & (y_pred == 1))
    tn = np.sum((y_true == 0) & (y_pred == 0))
    fp = np.sum((y_true == 0) & (y_pred == 1))
    fn = np.sum((y_true == 1) & (y_pred == 0))

    # Metrics
    acc  = (tp + tn) / max((tp + tn + fp + fn), 1)
    prec = tp / max((tp + fp), 1)
    rec  = tp / max((tp + fn), 1)
    f1   = (2 * prec * rec / max((prec + rec), 1e-12)) if (prec + rec) > 0 else 0.0

    # Specificity (True Negative Rate)
    spec = tn / max((tn + fp), 1)

    # Balanced accuracy
    bal_acc = 0.5 * (rec + spec)

    metrics = {
        "n": len(y_true),
        "tp": tp, "tn": tn, "fp": fp, "fn": fn,
        "acc": acc, "bal_acc": bal_acc, "prec": prec, "rec": rec, "spec": spec,
        "f1": f1
    }
    return metrics

def roc_auc_from_probs(y_true, y_prob):
    
    desc_sort_indices = np.argsort(-y_prob)
    y_true = np.array(y_true)[desc_sort_indices]
    y_prob = np.array(y_prob)[desc_sort_indices]
    pos = np.sum(y_true == 1)
    neg = np.sum(y_true == 0)

    # running totals for TPR/FPR
    tpr = [0.0]
    fpr = [0.0]
    tp = fp = 0
    for i in range(len(y_true)):
        if y_true[i] == 1:
            tp += 1
        else:
            fp += 1
        tpr.append(tp / pos)
        fpr.append(fp / neg)

    # get auc
    auc = np.trapz(tpr, fpr)
    return auc

#### Cross-validation function

Note that in `cross_val.ipynb`, we assigned folds already for each type of splitting to maintain consistent comparison across modeling. So, there will be no explicit splitting in this file, we will just use the folds already created according to the type of split we want to use.

Also, while we will be tuning many xgboost parameters, some we will be set consistently for reproducability. These include:

- eval_metric='auc' : tells xgboost to evaluate performance based on ROC-AUC
- random_state=42 : due to random subsampling in tree building, we fix the random seed to get the same results every run
- n_jobs=-1 : uses all available CPU cores in parallel to speed up training
- tree_method='hist' : a histogram-based algorithm provided by xgboost that can be faster than the default exact method with similar performance
- scale_pos_weight = neg/pos : corrects for class imbalance by upweighting minority class so the model focuses more on them in training. we may change this ratio throughout testing, but it won't be a part of grid searches

In [146]:
def cv_xgb(data, feature_cols, target_col, params=None):
    
    if params == None:
        params = {}

    fold_metrics = []
    fold_preds = []
    for f in data.fold.unique():

        # split into train and test based on folds
        train = data[data.fold != f]
        test = data[data.fold == f]
        X_train, y_train = train[feature_cols], train[target_col]
        X_test, y_test = test[feature_cols], test[target_col]

        # calculate counts for class weighting
        pos = (y_train == 1).sum()
        neg = (y_train == 0).sum()
        balanced_weight = (neg / max(pos, 1)) * 0.5

        # fit model with specified params
        model = XGBClassifier(eval_metric='auc', 
                              random_state=42, 
                              n_jobs=-1, 
                              tree_method='hist',
                              scale_pos_weight=balanced_weight,
                              **params
                            )
        model.fit(X_train, y_train)
        
        # get predictions (probablities and decisions)
        y_prob = model.predict_proba(X_test)[:, 1]
        y_train_prob = model.predict_proba(X_train)[:,1] # make training predictions as well to assess overfit
        y_pred = (y_prob >= 0.5).astype(int)

        # calculate classification metrics from previously defined functions
        metrics = classification_metrics(y_test, y_pred)
        metrics['roc_auc'] = roc_auc_from_probs(y_test, y_prob)
        metrics['train_roc_auc'] = roc_auc_from_probs(y_train, y_train_prob)
        metrics['fold'] = int(f)

        # add to list of all fold metrics
        fold_metrics.append(metrics)

        # store raw fold predictions for later threshold tuning
        fold_preds.append(pd.DataFrame({
            "fold": f,
            "y_true": y_test.values,
            "y_prob": y_prob
        }))

    # combine fold predictions into one long DataFrame
    preds_df = pd.concat(fold_preds, ignore_index=True)

    # summary DataFrame for quick metrics view
    metrics_df = pd.DataFrame(fold_metrics).sort_values("fold").reset_index(drop=True)

    # return results in dataframe
    return metrics_df, preds_df

#### A grid-search function:

Tests every combination of hyperparameters -- very slow/inefficient, so consider size of grid.

In [147]:
def grid_search_xgb(data, feature_cols, target_col, param_grid):
    
    # get all possible combinations of parameters
    keys = list(param_grid.keys())
    combos = [dict(zip(keys, v)) for v in product(*param_grid.values())]

    # initialize stuff for tracking and results
    results = []
    total = len(combos)
    start = time.time()
    next_checkpoint = 5 
    best_roc_auc = 0
    best_params = None

    # evaluate every possible combo
    for i, params in enumerate(combos, 1):

        # run cross validation and store results
        metrics, preds = cv_xgb(data, feature_cols, target_col, params)
        mean_roc_auc = metrics["roc_auc"].mean()

        results.append({
            'params': params,
            'mean_roc_auc': mean_roc_auc,
            'mean_f1': metrics['f1'].mean(),
            'mean_acc': metrics['acc'].mean(),
            'mean_bal_acc': metrics['bal_acc'].mean(),
            'mean_prec': metrics['prec'].mean(),
            'mean_rec': metrics['rec'].mean(),
        })

        # tracker for updates
        if  mean_roc_auc > best_roc_auc:
            best_roc_auc =  mean_roc_auc
            best_params = params

        # print progress checkpoints
        pct_done = (i/total)*100
        elapsed = time.time() - start
        if pct_done >= next_checkpoint or i == total:
            print(f"{i}/{total} ({pct_done:5.1f}% in {elapsed/60:.1f} mins) | Best ROC-AUC: {best_roc_auc:.4f} | Best Params: {best_params}")
            next_checkpoint += 5


    
    return pd.DataFrame(results).sort_values("mean_roc_auc", ascending=False).reset_index(drop=True)

# Model Development

**Notes:** 
- All evaluation will focus on stratified cross-validation, but we will test the other methods as well. 
- Recall that folds have been pre-assigned to ensure consistency across different model development processes
- For our other models, we have decided to scale + PCA, but this is not necessary for nonlinear tree-based algorithms like XGBoost
    - these can only really hurt XGBoost, so we will not use it here

**Process:**
1. Setting a baseline
    - evaluating an xgb model with all default parameters to build off of
2. Hyperparameter tuning
    - evaluate many different combinations of parameters
    - choose the best set based on average ROC-AUC across all folds
3. Holdout evaluation
    - evaluate on the corresponding holdout table. the performance here is what we will compare with other models (LR, SVC, LDA)
4. Threshold tuning
    - tweak the threshold on the best model to maximize another chosen metric (recall, precision, f1, balanced accuracy, etc.) 
        - note that roc-auc is not affected by threshold, hence the need a different optimizing metric
    - what metric we choose to optimize with threshold depends on business needs
        - consider the cost of mislabeling someone as high risk? or trusting an applicant that you shouldn't? will there be human review?
        - something we can include in the right up as optionality moving forward, not something we have to decide now on our own
        - "our model is very solid at ranking applicants from low-risk to high-risk, but in terms of actual classification, we can move the threshold based on what matters most to the business"
5. Fairness evaluation

## 0. Setup

Recall that we have different datasets for each type of cross validation. We are currently focusing on the stratified splitting method, but random and multiple-stratification methods are avaiable for testing/comparison.

In [148]:
apps_cv_strat = pd.read_csv("data/apps_cv_strat.csv")
apps_holdout_strat = pd.read_csv("data/apps_holdout_strat.csv")
target_col = 'TARGET'
feature_cols = [col for col in apps_cv_strat.columns if col not in 
                [target_col, 'SK_ID_CURR', 'fold', 'neighbors_target_mean_500', 'AGE_INT', 'CODE_GENDER_M']]

In [143]:
feature_cols

['CNT_CHILDREN',
 'AMT_INCOME_TOTAL',
 'AMT_CREDIT',
 'AMT_ANNUITY',
 'AMT_GOODS_PRICE',
 'REGION_POPULATION_RELATIVE',
 'DAYS_BIRTH',
 'DAYS_EMPLOYED',
 'DAYS_REGISTRATION',
 'DAYS_ID_PUBLISH',
 'OWN_CAR_AGE',
 'FLAG_MOBIL',
 'FLAG_EMP_PHONE',
 'FLAG_WORK_PHONE',
 'FLAG_CONT_MOBILE',
 'FLAG_PHONE',
 'FLAG_EMAIL',
 'CNT_FAM_MEMBERS',
 'REGION_RATING_CLIENT',
 'REGION_RATING_CLIENT_W_CITY',
 'HOUR_APPR_PROCESS_START',
 'REG_REGION_NOT_LIVE_REGION',
 'REG_REGION_NOT_WORK_REGION',
 'LIVE_REGION_NOT_WORK_REGION',
 'REG_CITY_NOT_LIVE_CITY',
 'REG_CITY_NOT_WORK_CITY',
 'LIVE_CITY_NOT_WORK_CITY',
 'EXT_SOURCE_1',
 'EXT_SOURCE_2',
 'EXT_SOURCE_3',
 'OBS_30_CNT_SOCIAL_CIRCLE',
 'DEF_30_CNT_SOCIAL_CIRCLE',
 'OBS_60_CNT_SOCIAL_CIRCLE',
 'DEF_60_CNT_SOCIAL_CIRCLE',
 'DAYS_LAST_PHONE_CHANGE',
 'FLAG_DOCUMENT_2',
 'FLAG_DOCUMENT_3',
 'FLAG_DOCUMENT_4',
 'FLAG_DOCUMENT_5',
 'FLAG_DOCUMENT_6',
 'FLAG_DOCUMENT_7',
 'FLAG_DOCUMENT_8',
 'FLAG_DOCUMENT_9',
 'FLAG_DOCUMENT_10',
 'FLAG_DOCUMENT_11',
 'FLAG_

Even though we aren't doing PCA and xgboost handles correlated/unecessary features well, we can still simply our model a bit by removing some of them. It may help us run faster and assess feature importance at the end. This may or may not be used depending on results, but I'll make the set now to keep as an option.

In [6]:
corr = apps_cv_strat[feature_cols].corr().abs()
upper = corr.where(np.triu(np.ones(corr.shape), k=1).astype(bool))
to_drop = [column for column in upper.columns if any(upper[column] > 0.95)]

print(f"Dropping {len(to_drop)} highly correlated features")

feature_cols_pruned = [f for f in feature_cols if f not in to_drop]

Dropping 35 highly correlated features


## 1. Setting a Baseline

Fitting an XGBoost model with default parameters to understand baseline predictive power and what we can build on.

In [80]:
baseline_metrics, baseline_preds = cv_xgb(apps_cv_strat, feature_cols, target_col, params=None) # send no parameters
baseline_metrics

n    tp     tn    fp    fn       acc   bal_acc      prec       rec  \
0  49156  1590  40661  4526  2379  0.859529  0.650222  0.259974  0.400605   
1  49156  1595  40446  4741  2374  0.855257  0.648472  0.251736  0.401864   
2  49156  1552  40520  4667  2417  0.855887  0.643874  0.249558  0.391030   
3  49155  1601  40634  4552  2368  0.859221  0.651318  0.260198  0.403376   
4  49154  1597  40725  4461  2371  0.861008  0.651872  0.263618  0.402470   

       spec        f1   roc_auc  train_roc_auc  fold  
0  0.899838  0.315320  0.765036       0.905793     1  
1  0.895080  0.309558  0.762482       0.907356     2  
2  0.896718  0.304672  0.762064       0.902676     3  
3  0.899261  0.316341  0.767256       0.903416     4  
4  0.901275  0.318572  0.765378       0.906415     5

Precision, recall, and F1 are all quite low, while ROC-AUC is relatively strong. This indicates that the model is doing a good job ranking applicants from low-risk to high-risk, as in assigning higher probabilities to true positives, but the actual 0/1 decisions at the default threshold of 0.5 are poor (likely due to class imbalance). Also, the training ROC-AUC is even higher than the baseline ROC-AUC indicating some overfit risk. 

Therefore, we expect significant improvement after threshold tuning at the end. This is also why we focus on ROC-AUC during hyperparameter tuning -- as long as the model’s ability to rank cases is good (high ROC-AUC), we can later adjust the threshold to manipulate precision/recall/F1.

## 2. Hyperparameter Tuning

Now, finding hyperparameters that maximize ROC-AUC (while keeping an eye on other metrics). 

XGBoost has a lot of parameters, many of which make a big impact on predictions, so the method of selecting the best ones is more complicated than the other linear models. A simple grid search over a huge parameter grid will take forever, considering just one cross-validated iteration of the baseline model took 15+ seconds. 

Therefore, a smart approach may be to do a multi-step grid search. There are two ways we could do this:

1. Tree structure --> learning dynamics
    - first a grid search on parameters that affect the shape of the trees essentially (max_depth, subsample, etc.)
    - then a grid search on paramaters that affect how the model learns (learning_rate, estimators, etc.)
2. Wide range, big step size --> small range, small step size
    - a grid search on *all* parameters with a very wide range for each parameter
    - then a more granular search on the best parameters to find exact optimal values

We will start with the first structural approach, but then may incorporate the second approach at some points as well. 

### First Pass:

Optimizing tree structure.

In [32]:
param_grid = {

    # fix learning dynamics this pass
    "learning_rate": [0.05], # low learning rate at first to establish reliable tree structure
    "n_estimators": [500], 

    # tree structure parameters to test
    "max_depth": [3, 4, 5, 6, 7],
    "min_child_weight": [1, 3, 5, 8, 12],
    "subsample": [0.7, 0.85, 1.0],
    "colsample_bytree": [0.7, 0.85, 1.0]
}

search1_results = grid_search_xgb(apps_cv_strat, feature_cols, target_col, param_grid)

12/225 (  5.3% in 7.5 mins) | Best ROC-AUC: 0.7790 | Best Params: {'learning_rate': 0.05, 'n_estimators': 500, 'max_depth': 3, 'min_child_weight': 3, 'subsample': 0.7, 'colsample_bytree': 0.85}
23/225 ( 10.2% in 14.7 mins) | Best ROC-AUC: 0.7790 | Best Params: {'learning_rate': 0.05, 'n_estimators': 500, 'max_depth': 3, 'min_child_weight': 3, 'subsample': 0.7, 'colsample_bytree': 0.85}
34/225 ( 15.1% in 21.5 mins) | Best ROC-AUC: 0.7791 | Best Params: {'learning_rate': 0.05, 'n_estimators': 500, 'max_depth': 3, 'min_child_weight': 8, 'subsample': 0.7, 'colsample_bytree': 0.85}
45/225 ( 20.0% in 27.5 mins) | Best ROC-AUC: 0.7791 | Best Params: {'learning_rate': 0.05, 'n_estimators': 500, 'max_depth': 3, 'min_child_weight': 12, 'subsample': 0.7, 'colsample_bytree': 0.85}
57/225 ( 25.3% in 34.6 mins) | Best ROC-AUC: 0.7810 | Best Params: {'learning_rate': 0.05, 'n_estimators': 500, 'max_depth': 4, 'min_child_weight': 3, 'subsample': 0.7, 'colsample_bytree': 0.7}
68/225 ( 30.2% in 41.3 min

In [34]:
# save so we dont have to run it all again
search1_results.to_csv('results/xgb_search1_results.csv')

In [47]:
search1_results.sort_values(by='mean_roc_auc', ascending=False)[
    ['params', 'mean_roc_auc', 'mean_prec', 'mean_rec', 'mean_f1']].head(10)

params  mean_roc_auc  mean_prec  \
0  {'learning_rate': 0.05, 'n_estimators': 500, '...      0.781358   0.184567   
1  {'learning_rate': 0.05, 'n_estimators': 500, '...      0.781235   0.192193   
2  {'learning_rate': 0.05, 'n_estimators': 500, '...      0.781106   0.184413   
3  {'learning_rate': 0.05, 'n_estimators': 500, '...      0.781089   0.184827   
4  {'learning_rate': 0.05, 'n_estimators': 500, '...      0.781088   0.192360   
5  {'learning_rate': 0.05, 'n_estimators': 500, '...      0.781088   0.184615   
6  {'learning_rate': 0.05, 'n_estimators': 500, '...      0.781080   0.184300   
7  {'learning_rate': 0.05, 'n_estimators': 500, '...      0.781050   0.191841   
8  {'learning_rate': 0.05, 'n_estimators': 500, '...      0.781029   0.184236   
9  {'learning_rate': 0.05, 'n_estimators': 500, '...      0.781008   0.184761   

   mean_rec   mean_f1  
0  0.685850  0.290855  
1  0.661913  0.297880  
2  0.684237  0.290517  
3  0.684439  0.291051  
4  0.661510  0.298040  
5  0.684943  0.290832  
6  0.686051  0.290543  
7  0.664533  0.297723  
8  0.683632  0.290245  
9  0.684539  0.290977

It looks like we found decent lift (+0.02 ROC-AUC) from tuning the tree structure. We also virtually identical scores at the top, not just one outlier, which is a good sign we found an optimal structure -- if these top scores had similar parameters, that would be further proof:

In [58]:
params_df = search1_results['params'].apply(pd.Series)
results = pd.concat([search1_results.drop(columns='params'), params_df], axis=1)
results.sort_values(by='mean_roc_auc', ascending=False)[
    ['mean_roc_auc', 'max_depth', 'min_child_weight', 'subsample', 'colsample_bytree']].head()

mean_roc_auc  max_depth  min_child_weight  subsample  colsample_bytree
0      0.781358        4.0               8.0       0.70               0.7
1      0.781235        5.0               8.0       0.85               1.0
2      0.781106        4.0              12.0       0.70               1.0
3      0.781089        4.0              12.0       0.85               1.0
4      0.781088        5.0               8.0       0.70               1.0

The top 5 models by ROC-AUC have very small changes in parameters -- The optimal region occurs with moderate `depth` (4–5), high `min_child_weight` (8–12), and subsampling between 0.7–0.85, suggesting strong generalization, reproducability, and minimal overfitting risk. We also don't see any sitting on a single edge, like if all the best performers were at the highest depth, so there is no reason to believe we are missing anything beyond our grid. 

These parameters will be fixed while tuning learning dynamics in the next phase:

- max_depth = 4
- min_child_weight = 8
- subsample = 0.7
- colsample_bytree = 1

These were the parameters of the best model by ROC-AUC, *except* for `colsample_bytree` because four out of the top five models had this parameter set to 1, so I felt it was safest to keep it as that, even though the #1 model was 0.7.

### Second Pass: Learning Dynamics

Now that we have a reliable tree structure, we will focus on the parameters that effect how the model learns. 

I think in this case it makes more sense to do the wide-then-narrow search approach because the learning parameters are continuous and small changes can make a big difference. Therefore, we will first find the appropriate scale of each parameter, and then do a narrow search over a smaller range to find more exact optimal parameters. Also, the difference from the best to 2nd (or even 5th) best is so small it's negligble.

**Finding Scale of Learning Dynamics:**

Bigger grid than before, so will take much longer. A reasonable option is to make 'reasonable' learning_rate + n_estimator pairings -- low learnings rates are very likely to underfit with a low number of estimators (trees) because it can't converge fast enough, so you can consider not running some of pairings you think are 'unreasonable'. 

However, I'm going to run this overnight either way and hopefully won't need to run again, so why not be thorough with the full grid. 

In [59]:
param_grid = {

    # fix tree structure to optimal parameters found in pass 1
    "max_depth": [4],
    "min_child_weight": [8],
    "subsample": [0.7],
    "colsample_bytree": [1],

    # wide grid to find scale of learning dynamics
    "learning_rate": [0.01, 0.03, 0.05, 0.07, 0.10],
    "n_estimators": [300, 500, 800, 1200],
    "gamma": [0.0, 0.1, 0.2, 0.3],
    "reg_alpha": [0.0, 0.5, 1.0],
    "reg_lambda": [1.0, 2.0, 5.0]
}

search2_results = grid_search_xgb(apps_cv_strat, feature_cols, target_col, param_grid)

36/720 (  5.0% in 20.0 mins) | Best ROC-AUC: 0.7565 | Best Params: {'max_depth': 4, 'min_child_weight': 8, 'subsample': 0.7, 'colsample_bytree': 1, 'learning_rate': 0.01, 'n_estimators': 300, 'gamma': 0.0, 'reg_alpha': 0.0, 'reg_lambda': 1.0}
72/720 ( 10.0% in 46.7 mins) | Best ROC-AUC: 0.7660 | Best Params: {'max_depth': 4, 'min_child_weight': 8, 'subsample': 0.7, 'colsample_bytree': 1, 'learning_rate': 0.01, 'n_estimators': 500, 'gamma': 0.0, 'reg_alpha': 0.5, 'reg_lambda': 1.0}
108/720 ( 15.0% in 82.6 mins) | Best ROC-AUC: 0.7727 | Best Params: {'max_depth': 4, 'min_child_weight': 8, 'subsample': 0.7, 'colsample_bytree': 1, 'learning_rate': 0.01, 'n_estimators': 800, 'gamma': 0.3, 'reg_alpha': 0.5, 'reg_lambda': 1.0}
144/720 ( 20.0% in 134.2 mins) | Best ROC-AUC: 0.7772 | Best Params: {'max_depth': 4, 'min_child_weight': 8, 'subsample': 0.7, 'colsample_bytree': 1, 'learning_rate': 0.01, 'n_estimators': 1200, 'gamma': 0.0, 'reg_alpha': 0.5, 'reg_lambda': 5.0}
180/720 ( 25.0% in 149.6

In [60]:
# save so we dont have to run it all again
search2_results.to_csv('results/xgb_search2_results.csv')

In [61]:
params_df = search2_results['params'].apply(pd.Series)
results = pd.concat([search2_results.drop(columns='params'), params_df], axis=1)
results.sort_values(by='mean_roc_auc', ascending=False)[
    ['mean_roc_auc', 'learning_rate', 'n_estimators', 'gamma', 'reg_alpha', 'reg_lambda']].head(10)

mean_roc_auc  learning_rate  n_estimators  gamma  reg_alpha  reg_lambda
0      0.783057           0.03        1200.0    0.0        0.5         5.0
1      0.782981           0.03        1200.0    0.3        0.5         5.0
2      0.782968           0.03        1200.0    0.1        0.5         5.0
3      0.782923           0.03        1200.0    0.2        0.5         5.0
4      0.782863           0.05         800.0    0.1        0.0         5.0
5      0.782842           0.03        1200.0    0.2        0.5         1.0
6      0.782842           0.05         800.0    0.0        0.0         5.0
7      0.782786           0.03        1200.0    0.3        0.5         1.0
8      0.782783           0.05         800.0    0.2        0.0         5.0
9      0.782778           0.05         800.0    0.3        0.0         5.0

We see some lift again (+0.002) ROC-AUC, but much smaller than the last pass, which is expected, as we were just hoping to squeeze out points at this stage. Additionally, since all of the top scorers have almost identical performance, we can assume we are close to optimal learning dynamics.

However, I notice that `n_estimators` and `reg_lambda` are consistently on an edge of our parameter range, so before doing our narrow search, I want to expand the scale of these a bit to see if they continue to operate better at higher scale. Also, `gamma` does not seem to have an optimal value at all, so I'll test a wider range for that as well, but suspecting it may not have a real impact on our trees. 

Note that increasing `n_estimators` has the biggest impact on speed, so that is something to consider if speed is required for this model -- but for now, we are just looking for best performance.

In [62]:
param_grid = {

    # fix tree structure to optimal parameters found in pass 1
    "max_depth": [4],
    "min_child_weight": [8],
    "subsample": [0.7],
    "colsample_bytree": [1],

    "learning_rate": [0.03], # fixed, stable
    "n_estimators": [1200, 1500, 1800, 2000], # expanding higher
    "gamma": [0.0, 0.2, 0.4, 0.6], # trying some other values 
    "reg_alpha": [0.5], # fixed, stable
    "reg_lambda": [5.0, 7.5, 10.0, 15.0] # expanding higher
}

search2_expanded_results = grid_search_xgb(apps_cv_strat, feature_cols, target_col, param_grid)

4/64 (  6.2% in 6.6 mins) | Best ROC-AUC: 0.7833 | Best Params: {'max_depth': 4, 'min_child_weight': 8, 'subsample': 0.7, 'colsample_bytree': 1, 'learning_rate': 0.03, 'n_estimators': 1200, 'gamma': 0.0, 'reg_alpha': 0.5, 'reg_lambda': 15.0}
7/64 ( 10.9% in 10.6 mins) | Best ROC-AUC: 0.7833 | Best Params: {'max_depth': 4, 'min_child_weight': 8, 'subsample': 0.7, 'colsample_bytree': 1, 'learning_rate': 0.03, 'n_estimators': 1200, 'gamma': 0.0, 'reg_alpha': 0.5, 'reg_lambda': 15.0}
10/64 ( 15.6% in 14.9 mins) | Best ROC-AUC: 0.7833 | Best Params: {'max_depth': 4, 'min_child_weight': 8, 'subsample': 0.7, 'colsample_bytree': 1, 'learning_rate': 0.03, 'n_estimators': 1200, 'gamma': 0.2, 'reg_alpha': 0.5, 'reg_lambda': 15.0}
13/64 ( 20.3% in 19.0 mins) | Best ROC-AUC: 0.7834 | Best Params: {'max_depth': 4, 'min_child_weight': 8, 'subsample': 0.7, 'colsample_bytree': 1, 'learning_rate': 0.03, 'n_estimators': 1200, 'gamma': 0.4, 'reg_alpha': 0.5, 'reg_lambda': 15.0}
16/64 ( 25.0% in 23.0 mins)

In [81]:
# save so we dont have to run it all again
search2_expanded_results.to_csv('results/xgb_search2_expanded_results.csv')

In [64]:
params_df = search2_expanded_results['params'].apply(pd.Series)
results = pd.concat([search2_expanded_results.drop(columns='params'), params_df], axis=1)
results.sort_values(by='mean_roc_auc', ascending=False)[
    ['mean_roc_auc', 'n_estimators', 'gamma', 'reg_lambda']].head(10)

mean_roc_auc  n_estimators  gamma  reg_lambda
0      0.783527        1500.0    0.6        15.0
1      0.783494        1500.0    0.4        15.0
2      0.783410        1500.0    0.0        15.0
3      0.783397        1500.0    0.2        15.0
4      0.783394        1200.0    0.4        15.0
5      0.783341        1200.0    0.6        15.0
6      0.783317        1800.0    0.4        15.0
7      0.783302        1200.0    0.2        15.0
8      0.783298        1800.0    0.6        15.0
9      0.783287        1500.0    0.0        10.0

- `n_estimators` seems optimal at 1500
- `gamma` appears to not affect the model. probably because we have so much regularization elsewhere that it doesn't matter much
- `reg_lambda` is still at the top of our range so lets do one final sweep to see if any more performance can be squeezed out before diminishing returns

A final search for `reg_lambda` scale before narrowing all parameters further:

In [75]:
param_grid = {

    # fix tree structure to optimal parameters found in pass 1
    "max_depth": [4],
    "min_child_weight": [8],
    "subsample": [0.7],
    "colsample_bytree": [1],

    "learning_rate": [0.03], # fixed, stable
    "n_estimators": [1500], # fixed, stable
    "gamma": [1], # fixed, not very useful
    "reg_alpha": [0.5], # fixed, stable
    "reg_lambda": [10, 15, 20, 25, 30] # further expansion
}

reg_lambda_expansion = grid_search_xgb(apps_cv_strat, feature_cols, target_col, param_grid)

1/5 ( 20.0% in 2.2 mins) | Best ROC-AUC: 0.7831 | Best Params: {'max_depth': 4, 'min_child_weight': 8, 'subsample': 0.7, 'colsample_bytree': 1, 'learning_rate': 0.03, 'n_estimators': 1500, 'gamma': 1, 'reg_alpha': 0.5, 'reg_lambda': 10}
2/5 ( 40.0% in 4.0 mins) | Best ROC-AUC: 0.7836 | Best Params: {'max_depth': 4, 'min_child_weight': 8, 'subsample': 0.7, 'colsample_bytree': 1, 'learning_rate': 0.03, 'n_estimators': 1500, 'gamma': 1, 'reg_alpha': 0.5, 'reg_lambda': 15}
3/5 ( 60.0% in 5.7 mins) | Best ROC-AUC: 0.7836 | Best Params: {'max_depth': 4, 'min_child_weight': 8, 'subsample': 0.7, 'colsample_bytree': 1, 'learning_rate': 0.03, 'n_estimators': 1500, 'gamma': 1, 'reg_alpha': 0.5, 'reg_lambda': 15}
4/5 ( 80.0% in 7.4 mins) | Best ROC-AUC: 0.7836 | Best Params: {'max_depth': 4, 'min_child_weight': 8, 'subsample': 0.7, 'colsample_bytree': 1, 'learning_rate': 0.03, 'n_estimators': 1500, 'gamma': 1, 'reg_alpha': 0.5, 'reg_lambda': 15}
5/5 (100.0% in 9.2 mins) | Best ROC-AUC: 0.7836 | Be

In [ ]:
# save so we dont have to run it all again
reg_lambda_expansion.to_csv('results/xgb_reg_lambda_expansion_results.csv')

In [76]:
params_df = reg_lambda_expansion['params'].apply(pd.Series)
results = pd.concat([reg_lambda_expansion.drop(columns='params'), params_df], axis=1)
results.sort_values(by='mean_roc_auc', ascending=False)[
    ['mean_roc_auc', 'reg_lambda']].head(10)

mean_roc_auc  reg_lambda
0      0.783609        15.0
1      0.783512        25.0
2      0.783464        20.0
3      0.783450        30.0
4      0.783123        10.0

It looks like we do peak at `reg_lambda`=15 before flattening out. 

I think we have found the appropriate scale for each of our learning dynamics hyperparameters:

- `learning_rate` = 0.03
- `n_estimators` = 1500
- `gamma` = 1
- `reg_alpha` = 0.5
- `reg_lambda` = 15

Now, a refined, narrow search around these parameters. Not expecting a huge lift here.

**Finding exact optimal parameters:**

We have already determined that `gamma` does not affect much and that `reg_lambda` pleateaus hard at 15, so these do not need to be further fine-tuned. Changes due to `n_estimtors` tend to be pretty smooth, and we already tested with a step size of 300, so we can't expect too much of a difference, but a smaller step size could be worth trying. `learning_rate` and `reg_alpha` will be the primary parameters to refine.

In [82]:
param_grid = {

    # fix tree structure to optimal parameters found in pass 1
    "max_depth": [4],
    "min_child_weight": [8],
    "subsample": [0.7],
    "colsample_bytree": [1],

    # very narrow search over the parameters we just found
    "learning_rate": [0.02, 0.03, 0.04], 
    "n_estimators": [1350, 1500, 1650], 
    "gamma": [1], # fixed, not very useful
    "reg_alpha": [0.3, 0.4, 0.5, 0.6, 0.7],
    "reg_lambda": [15] 
}

narrow_learning_search = grid_search_xgb(apps_cv_strat, feature_cols, target_col, param_grid)

3/45 (  6.7% in 5.5 mins) | Best ROC-AUC: 0.7828 | Best Params: {'max_depth': 4, 'min_child_weight': 8, 'subsample': 0.7, 'colsample_bytree': 1, 'learning_rate': 0.02, 'n_estimators': 1350, 'gamma': 1, 'reg_alpha': 0.5, 'reg_lambda': 15}
5/45 ( 11.1% in 9.2 mins) | Best ROC-AUC: 0.7828 | Best Params: {'max_depth': 4, 'min_child_weight': 8, 'subsample': 0.7, 'colsample_bytree': 1, 'learning_rate': 0.02, 'n_estimators': 1350, 'gamma': 1, 'reg_alpha': 0.5, 'reg_lambda': 15}
7/45 ( 15.6% in 12.7 mins) | Best ROC-AUC: 0.7832 | Best Params: {'max_depth': 4, 'min_child_weight': 8, 'subsample': 0.7, 'colsample_bytree': 1, 'learning_rate': 0.02, 'n_estimators': 1500, 'gamma': 1, 'reg_alpha': 0.4, 'reg_lambda': 15}
9/45 ( 20.0% in 16.3 mins) | Best ROC-AUC: 0.7832 | Best Params: {'max_depth': 4, 'min_child_weight': 8, 'subsample': 0.7, 'colsample_bytree': 1, 'learning_rate': 0.02, 'n_estimators': 1500, 'gamma': 1, 'reg_alpha': 0.5, 'reg_lambda': 15}
12/45 ( 26.7% in 21.6 mins) | Best ROC-AUC: 0.

In [83]:
# save so we dont have to run it all again
narrow_learning_search.to_csv('results/xgb_narrow_learning_search_results.csv')

In [85]:
params_df = narrow_learning_search['params'].apply(pd.Series)
results = pd.concat([narrow_learning_search.drop(columns='params'), params_df], axis=1)
results.sort_values(by='mean_roc_auc', ascending=False)[
    ['mean_roc_auc', 'learning_rate', 'n_estimators', 'reg_alpha']].head(10)

mean_roc_auc  learning_rate  n_estimators  reg_alpha
0      0.783621           0.03        1350.0        0.5
1      0.783609           0.03        1500.0        0.5
2      0.783561           0.03        1650.0        0.5
3      0.783549           0.03        1350.0        0.3
4      0.783547           0.03        1500.0        0.3
5      0.783544           0.03        1500.0        0.7
6      0.783534           0.03        1350.0        0.7
7      0.783519           0.02        1650.0        0.5
8      0.783482           0.03        1500.0        0.6
9      0.783476           0.03        1350.0        0.6

### Confirmation

Now I just want to confirm that these results are stable, not overfitting, and safe from leakage. 

**Stability across folds:**

In [81]:
tuned_params = {

    # tree structure
    "max_depth": 4,
    "min_child_weight": 8,
    "subsample": 0.7,
    "colsample_bytree": 1,

    # learning dynamics
    "learning_rate": 0.03, 
    "n_estimators": 1350, 
    "gamma": 1,
    "reg_alpha": 0.5,
    "reg_lambda": 15 

    # rest are already set in cv function
}

fold_metrics, preds = cv_xgb(apps_cv_strat, feature_cols, target_col, tuned_params)
fold_metrics

n    tp     tn    fp    fn       acc   bal_acc      prec       rec  \
0  49156  1745  40303  4884  2224  0.855399  0.665787  0.263237  0.439657   
1  49156  1789  40128  5059  2180  0.852734  0.669393  0.261244  0.450743   
2  49156  1732  40293  4894  2237  0.854931  0.664038  0.261395  0.436382   
3  49155  1760  40374  4812  2209  0.857166  0.668472  0.267803  0.443437   
4  49154  1757  40407  4779  2211  0.857794  0.668515  0.268819  0.442792   

       spec        f1   roc_auc  train_roc_auc  fold  
0  0.891916  0.329307  0.781825       0.843168     1  
1  0.888043  0.330776  0.779880       0.844188     2  
2  0.891695  0.326947  0.780157       0.844102     3  
3  0.893507  0.333934  0.783474       0.842740     4  
4  0.894237  0.334539  0.782210       0.842628     5

In [82]:
print(f"Mean AUC: {fold_metrics.roc_auc.mean():.6f}")
print(f"Std AUC: {fold_metrics.roc_auc.std():.6f}")

Mean AUC: 0.781509
Std AUC: 0.001494


Appears to be extremely stable. Very small differences across folds, meaning that the average score isn't carried by just one fold getting 'lucky'. We can assume consistent performance from our model. 

**Overfit check:**

Want to see a minimal gap between trainings and testing ROC-AUC

In [83]:
print("Mean Train AUC:", fold_metrics.train_roc_auc.mean())
print("Mean Val AUC:", fold_metrics.roc_auc.mean())
print("Gap:", fold_metrics.train_roc_auc.mean() - fold_metrics.roc_auc.mean())

Mean Train AUC: 0.8433651981349026
Mean Val AUC: 0.7815091704034669
Gap: 0.061856027731435725


There is definitely a little bit of overfit, but the stability across folds also mitigates this risk. We can try adding some regularization / making the tree less complex to minimize that gap while keeping performance high.

In [84]:
final_params = {

    # tree structure
    "max_depth": 4,
    "min_child_weight": 8,
    "subsample": 0.5,
    "colsample_bytree": 0.7,

    # learning dynamics
    "learning_rate": 0.03, 
    "n_estimators": 1000, 
    "gamma": 2,
    "reg_alpha": 0.5,
    "reg_lambda": 20 

    # rest are already set in cv function
}
fold_metrics, preds = cv_xgb(apps_cv_strat, feature_cols, target_col, final_params)
print("Mean Train AUC:", fold_metrics.train_roc_auc.mean())
print("Mean Val AUC:", fold_metrics.roc_auc.mean())
print("Gap:", fold_metrics.train_roc_auc.mean() - fold_metrics.roc_auc.mean())
print(f"Std AUC: {fold_metrics.roc_auc.std():.6f}")

Mean Train AUC: 0.8262172177839059
Mean Val AUC: 0.7808552387120334
Gap: 0.04536197907187245
Std AUC: 0.001646


After making a few tweaks to the complexity of the model, we have closed the gap between training and validation performance by 26% without affecting validation performance in a significant way. I believe the gap is now within a comfortable range to assume we are *not* overfitting. Variance between folds even decreased a bit too. 

**Leakage Check:**

Lastly, I want to confirm that we have a *safe* model by confirming as best as possible that we evaluated our model properly, especially by not revealing information from our test set to our training set.

One way to assess this is by shuffling the target variable, removing any structural patterns from the data, which means we should expect ~0.5 AUC (randomly guessing). If we see significantly higher, it would indicate that the model is exploiting some sort of pattern that it could only pick up on through data leaking from testing to training.   

In [86]:
shuffled = apps_cv_strat.copy()
shuffled['TARGET'] = np.random.permutation(shuffled['TARGET'].values)
fold_metrics_shuffled, preds_shuffled = cv_xgb(shuffled, feature_cols, target_col, final_params)
print("Shuffled mean AUC:", fold_metrics_shuffled.roc_auc.mean())

Shuffled mean AUC: 0.5001990094286135


Perfect, no issues.

### Final Hyperparameters

Now that we have confirmed a safe and reliabel model, reducing risk of overfitting and leakage, I think we are comfortable with these hyperparameters (including the pre-set params):

- `max_depth` = 4
- `min_child_weight` = 8
- `subsample` = 0.5
- `colsample_bytree` = 0.7
- `learning_rate` = 0.03
- `n_estimators` = 1000
- `gamma` = 2
- `reg_alpha` = 0.5
- `reg_lambda` = 20
- `scale_pos_weight` = 0.5 * #neg/#pos (found the 0.5 scaler helped a little)
- `eval_metric` = 'auc'
- `tree_method` = 'hist'
- `random_state` = 42
- `n_jobs` = -1

# 3. Holdout Evaluation

We are going to compare this model to other models through performance on the holdout set. To do this, we will train on the set model above on the entire dataset we have been using so far for model selection, and then test it on the entire holdout set that we have not seen yet for unbiased evaluation. 

In [108]:
# separate data
X_train, y_train = apps_cv_strat[feature_cols], apps_cv_strat[target_col]
X_test, y_test = apps_holdout_strat[feature_cols], apps_holdout_strat[target_col]

# initiliaze our finalized parameters
pos = (y_train == 1).sum()
neg = (y_train == 0).sum()
params = {
    "max_depth": 4,
    "min_child_weight": 8,
    "subsample": 0.5,
    "colsample_bytree": 0.7,
    "learning_rate": 0.03, 
    "n_estimators": 1000, 
    "gamma": 2,
    "reg_alpha": 0.5,
    "reg_lambda": 20,
    'eval_metric': 'auc', 
    'random_state': 42, 
    'n_jobs': -1, 
    'tree_method': 'hist',
    'scale_pos_weight': (neg / max(pos, 1)) * 0.5 
}

# fit model on training data
model = XGBClassifier(**params)
model.fit(X_train, y_train)

# predict on holdout data
y_prob = model.predict_proba(X_test)[:, 1]
y_pred  = (y_prob >= 0.5).astype(int) # just the default 0.5 threshold for now
y_train_prob = model.predict_proba(X_train)[:,1] # make training predictions as well to assess overfit

# calculate classification metrics from previously defined functions
metrics = classification_metrics(y_test, y_pred)
metrics['roc_auc'] = roc_auc_from_probs(y_test, y_prob)
metrics['train_roc_auc'] = roc_auc_from_probs(y_train, y_train_prob)
metrics

{'n': 61443,
 'tp': 2302,
 'tn': 50180,
 'fp': 6303,
 'fn': 2658,
 'acc': 0.8541575118402421,
 'bal_acc': 0.6762609025096332,
 'prec': 0.26751888436955257,
 'rec': 0.46411290322580645,
 'spec': 0.88840890179346,
 'f1': 0.3394028750460744,
 'roc_auc': 0.782289486331314,
 'train_roc_auc': 0.81890841418022}

Since we still haven't tuned threshold, we are really only looking at ROC-AUC right now. 

The performance from cross-validation carried over to the holdout set (with small improvement even), indicating we have a very generalizable model with no leakage risks. Overfitting is still at low-risk. Overall, a successful run on the holdout set, and I'm pretty confident we squeezed out as much performance as we can from this XGBoost model.

An `ROC-AUC` of **0.784** is what we will compare to our other models.

### Kaggle Predictions

The Kaggle competition has it's own holdout set that we need to predict on for our submission. This doesn't have the true targets so we can't evaluate it ourselves, but we can make the predictions and submit it for an overall score. 

In [88]:
# read in the full training and test data
## AFTER running through our pipeline files
train = pd.read_csv('data/apps_all_background.csv')
test = pd.read_csv('data/apps_all_background_test.csv')

# separate data
X_train, y_train = train[feature_cols], train[target_col]
X_test = test[feature_cols]

# fit model on training data
model = XGBClassifier(**params)
model.fit(X_train, y_train)

# predict on holdout data
test["TARGET"] = model.predict_proba(X_test)[:, 1]

# create submission csv
submission = test[["SK_ID_CURR", "TARGET"]]
submission.to_csv("results/submission_xgb.csv", index=False)

**RESULTS:** 
- `Public`: 0.788
- `Private`: 0.781

Another submission with the parameters *before* adjusting for overfit just to try:

In [89]:
params = {
    "max_depth": 4,
    "min_child_weight": 8,
    "subsample": 0.7,
    "colsample_bytree": 1,
    "learning_rate": 0.03, 
    "n_estimators": 1350, 
    "gamma": 1,
    "reg_alpha": 0.5,
    "reg_lambda": 15,
    'eval_metric': 'auc', 
    'random_state': 42, 
    'n_jobs': -1, 
    'tree_method': 'hist',
    'scale_pos_weight': (neg / max(pos, 1)) * 0.5 
}

# fit model on training data
model = XGBClassifier(**params)
model.fit(X_train, y_train)

# predict on holdout data
test["TARGET"] = model.predict_proba(X_test)[:, 1]

# create submission csv
submission = test[["SK_ID_CURR", "TARGET"]]
submission.to_csv("results/submission2_xgb.csv", index=False)

**RESULTS:** 
- `Public`: 0.790
- `Private`: 0.783

I'm going to test the xgboost model on just the given application data on our holdout set as an experiment so we can measure lift from feature engineering + lift from tuning:

In [98]:
# read in original applications table
apps = pd.read_csv("data/application_train.csv")  

# get id's from current tables we're using to matchup with original tables
train_prev = pd.read_csv("data/apps_cv_strat.csv", usecols=["SK_ID_CURR"])
hold_prev  = pd.read_csv("data/apps_holdout_strat.csv",   usecols=["SK_ID_CURR"])
train_ids = set(train_prev["SK_ID_CURR"])
hold_ids  = set(hold_prev["SK_ID_CURR"])

# split original applications data based on how we already split current data
apps_train = apps[apps["SK_ID_CURR"].isin(train_ids)].copy()
apps_hold  = apps[apps["SK_ID_CURR"].isin(hold_ids)].copy()

# add in folds for cross validation
folds = pd.read_csv("data/apps_cv_strat.csv", usecols=["SK_ID_CURR", "fold"])
apps_train = apps_train.merge(folds, on="SK_ID_CURR", how="left")

# cast objects to category (xgboost can handle it)
for df in (apps_train, apps_hold):
    for c in df.select_dtypes(include="object").columns:
        df[c] = df[c].astype("category")

target_col = "TARGET"
feature_cols_apps = [c for c in apps.columns if c not in ["SK_ID_CURR", target_col, "fold", 'CODE_GENDER_M', 'CODE_GENDER', 'AGE_INT']]
X_apps_train, y_apps_train = apps_train[feature_cols_apps], apps_train[target_col]

print("X_apps_train shape:", X_apps_train.shape)

X_apps_train shape: (245777, 119)


In [99]:
og_data_metrics, og_data_preds = cv_xgb(apps_train, feature_cols_apps, target_col, params={'enable_categorical':True})
og_data_metrics

n    tp     tn    fp    fn       acc   bal_acc      prec       rec  \
0  49156  1421  40661  4526  2548  0.856091  0.628932  0.238944  0.358025   
1  49156  1464  40431  4756  2505  0.852287  0.631804  0.235370  0.368859   
2  49156  1379  40619  4568  2590  0.854382  0.623176  0.231882  0.347443   
3  49155  1399  40753  4433  2570  0.857532  0.627188  0.239883  0.352482   
4  49154  1439  40543  4643  2529  0.854091  0.629949  0.236600  0.362651   

       spec        f1   roc_auc  train_roc_auc  fold  
0  0.899838  0.286608  0.735134       0.895877     1  
1  0.894748  0.287369  0.729318       0.898322     2  
2  0.898909  0.278136  0.731745       0.899228     3  
3  0.901894  0.285481  0.735326       0.898300     4  
4  0.897247  0.286368  0.734011       0.899578     5

So overall for our 3 main stages of improvement:

| Stage                            | Description                                                                          | Key Changes                                                                   |    Mean ROC-AUC   |              Approx. Lift from Base             |
| :------------------------------- | :----------------------------------------------------------------------------------- | :---------------------------------------------------------------------------- | :---------------: | :-----------------------------------: |
| **1. Base** | Model trained only on the raw `application_train.csv` features.                      | No joins or aggregations. Baseline feature set.                               | **~0.73** |                                      |
| **2. Feature-Engineered**       | Added engineered + aggregated variables from internal / external credit tables.      | Joins, aggregations, missing-value flags, structural imputations.                           | **~0.76** |               **+0.026**              |
| **3️. Tuned XGBoost**            | Final optimized parameters (depth = 4, min_child_weight = 8, subsample = 0.5, etc.). | Full cross-validated tuning on tree structure + learning dynamics. | **~0.784** | **+0.047** |


# 4. Threshold Tuning

With a solid ROC-AUC, we know that our model can rank applicants based on risk well, but the cutoff for decision-making is highly business-dependent. The main question that needs to be considered is the cost of mislabeling in both directions. We can attempt to answer this on our own and come up with a proposed solution, but it can also be something we leave a bit open-ended in our report asking for feedback on what the client would prefer. 

In [92]:
# try 200 thresholds evenly spaced from 0 to 1
## should be fast. no model required with the predictions we already have

thresholds = np.linspace(0, 1, 200)
scores = []
for t in thresholds:
    y_pred = (preds["y_prob"] >= t).astype(int)
    tp = ((y_pred == 1) & (preds["y_true"] == 1)).sum()
    fp = ((y_pred == 1) & (preds["y_true"] == 0)).sum()
    tn = ((y_pred == 0) & (preds["y_true"] == 0)).sum()
    fn = ((y_pred == 0) & (preds["y_true"] == 1)).sum()

    prec = tp / max(tp + fp, 1)
    rec  = tp / max(tp + fn, 1)
    f1   = 2 * prec * rec / max(prec + rec, 1e-12)
    spec = tn / max(tn + fp, 1)
    bal_acc = (rec + spec) / 2

    scores.append((t, prec, rec, f1, bal_acc))

scores_df = pd.DataFrame(scores, columns=["threshold", "precision", "recall", "f1", "bal_acc"])

Sorted by precision:

In [93]:
scores_df[scores_df.precision != 1].sort_values("precision", ascending=False).head(5)

threshold  precision    recall        f1   bal_acc
186   0.934673   0.944444  0.000857  0.001712  0.500426
185   0.929648   0.913043  0.001058  0.002114  0.500525
183   0.919598   0.803279  0.002469  0.004923  0.501208
184   0.924623   0.800000  0.001411  0.002817  0.500690
182   0.914573   0.746988  0.003124  0.006223  0.501516

Sorted by recall:

In [94]:
scores_df[scores_df.recall != 1].sort_values("recall", ascending=False).head(5)

threshold  precision    recall        f1   bal_acc
4   0.020101   0.080977  0.999798  0.149819  0.501590
5   0.025126   0.081274  0.999647  0.150325  0.503570
6   0.030151   0.081698  0.999496  0.151049  0.506376
7   0.035176   0.082248  0.998740  0.151980  0.509962
8   0.040201   0.082958  0.998085  0.153184  0.514516

Sorted by F1:

In [95]:
scores_df.sort_values("f1", ascending=False).head(5)

threshold  precision    recall        f1   bal_acc
102   0.512563   0.269483  0.428492  0.330875  0.663235
101   0.507538   0.266253  0.435043  0.330336  0.664871
100   0.502513   0.263471  0.442552  0.330299  0.666946
103   0.517588   0.271823  0.420077  0.330067  0.660619
99    0.497487   0.260439  0.450111  0.329959  0.668924

Sorted by balanced accuracy:

In [96]:
scores_df.sort_values("bal_acc", ascending=False).head(5)

threshold  precision    recall        f1   bal_acc
60   0.301508   0.169645  0.735235  0.275681  0.709577
59   0.296482   0.167674  0.742643  0.273579  0.709428
61   0.306533   0.171411  0.727071  0.277419  0.709189
63   0.316583   0.175462  0.712256  0.281562  0.709140
58   0.291457   0.165667  0.748992  0.271322  0.708843

Obviously there is a big tradeoff between precision and recall. For the rest of our evaluation, we will continue with continue with balanced accuracy as our optimizing metric, but will give other options as:

| Goal                                  | Choose threshold | Reason                                                                      |
| ------------------------------------- | ---------------- | ------------------------------------------------------------------------------ |
| **Balanced performance**              | ~ 0.30      | Maximizes Balanced Accuracy (~0.71) and keeps F1 moderate.                       |
| **Cautious - low false alarms**       | ≥ 0.45           | Higher precision, lower recall - good if rejecting loans is costly.            |
| **Aggressive - flag all defaulters** | ≤ 0.20           | Boosts recall but also false positives - good for an initial screen followed by human review. |


And anywhere inbetween these ranges to balance different objectives.

# 5. Fairness Evaluation

The groups from our data we are interested in are gender (male vs. female) and age (binned into decades). Note that these features have been removed from our model for *Group Unawarness*. Since unknown proxy attributes may exist, we will compute various metrics as confirmation.

After fixing the decision threshold at 0.3 (chosen to maximize balanced accuracy), we will evaluate whether the model treats these groups equitably using several fairness metrics from class:

- Demographic Parity: compares overall positive prediction rates across groups
- Equal Opportunity: compares true positive rates (recall) across groups
- Equalized Odds: compares both true and false positive rates across groups
- Predictive Parity (PPVP): compares precision across groups

These metrics will be computed on the holdout predictions to assess if any group is disproportionately disadvantaged by the model.

First, fit holdout data again and get demographic groups:

In [123]:
# separate data
X_train, y_train = apps_cv_strat[feature_cols], apps_cv_strat[target_col]
X_test, y_test = apps_holdout_strat[feature_cols], apps_holdout_strat[target_col]

# initiliaze our finalized parameters
pos = (y_train == 1).sum()
neg = (y_train == 0).sum()
params = {
    "max_depth": 4,
    "min_child_weight": 8,
    "subsample": 0.5,
    "colsample_bytree": 0.7,
    "learning_rate": 0.03, 
    "n_estimators": 1000, 
    "gamma": 2,
    "reg_alpha": 0.5,
    "reg_lambda": 20,
    'eval_metric': 'auc', 
    'random_state': 42, 
    'n_jobs': -1, 
    'tree_method': 'hist',
    'scale_pos_weight': (neg / max(pos, 1)) * 0.5 
}

# fit model on training data
model = XGBClassifier(**params)
model.fit(X_train, y_train)

# predict on holdout data
apps_holdout_strat['y_prob'] = model.predict_proba(X_test)[:, 1]
apps_holdout_strat['y_pred']  = (apps_holdout_strat['y_prob'] >= 0.3).astype(int) # optimal 0.3 threshold

In [132]:
# convert gender back to readable label
apps_holdout_strat['GENDER'] = np.where(apps_holdout_strat['CODE_GENDER_M'], 'Male', 'Female')

# bin ages by decades
bins   = [20,30,40,50,60,70]
labels = ["20–29","30–39","40–49","50–59","60–69"]
apps_holdout_strat['AGE_GROUP'] = pd.cut(apps_holdout_strat['AGE_INT'], bins=bins, labels=labels, right=False)

A function for fairness metrics:

In [140]:
# function to caluclate fairness metrics based on a protected group
def fairness_metrics(df, group_col, y_true="TARGET", y_pred="y_pred"):
    results = []
    for g, sub in df.groupby(group_col):
        tp = ((sub[y_pred]==1) & (sub[y_true]==1)).sum()
        fp = ((sub[y_pred]==1) & (sub[y_true]==0)).sum()
        tn = ((sub[y_pred]==0) & (sub[y_true]==0)).sum()
        fn = ((sub[y_pred]==0) & (sub[y_true]==1)).sum()
        
        # core rates
        pos_rate = (tp + fp) / max(len(sub), 1)
        tpr = tp / max(tp + fn, 1)   # recall
        fpr = fp / max(fp + tn, 1)
        prec = tp / max(tp + fp, 1)
        
        results.append({
            group_col: g,
            "n": len(sub),
            "Pos Rate": pos_rate,
            "TPR (Equal Opportunity)": tpr,
            "FPR": fpr,
            "Precision (PPVP)": prec,
        })

    results_df = pd.DataFrame(results)
    results_df['TPR_Ratio'] = results_df['TPR (Equal Opportunity)'] / results_df['TPR (Equal Opportunity)'].max()
    results_df['FPR_Ratio'] = results_df['FPR'] / results_df['FPR'].max()

    return results_df

In [141]:
fairness_metrics(apps_holdout_strat, 'GENDER', 'TARGET', 'y_pred')

GENDER      n  Pos Rate  TPR (Equal Opportunity)       FPR  \
0  Female  40409  0.329580                 0.719603  0.300309   
1    Male  21034  0.406152                 0.783544  0.363429   

   Precision (PPVP)  TPR_Ratio  FPR_Ratio  
0          0.152425   0.918395   0.826319  
1          0.196184   1.000000   1.000000

All metrics are higher for men, but all are relatively close. 

In [142]:
fairness_metrics(apps_holdout_strat, 'AGE_GROUP', 'TARGET', 'y_pred')

AGE_GROUP      n  Pos Rate  TPR (Equal Opportunity)       FPR  \
0     20–29   9056  0.553445                 0.848963  0.516225   
1     30–39  16454  0.430594                 0.818882  0.390011   
2     40–49  15284  0.326551                 0.717077  0.293968   
3     50–59  13465  0.253695                 0.634409  0.228461   
4     60–69   7184  0.188892                 0.521277  0.170535   

   Precision (PPVP)  TPR_Ratio  FPR_Ratio  
0          0.171588   1.000000   1.000000  
1          0.179958   0.964567   0.755506  
2          0.169104   0.844650   0.569456  
3          0.155445   0.747274   0.442560  
4          0.144436   0.614015   0.330349